ambry-router/src/main/java/com.github.ambry.router/SimpleOperationTracker.java

```diff
@@ -227,22 +240,32 @@
         replicaPool.add(downReplicas.pollFirst());
       }
     }
+    cloudReplicasPresent =
+        replicaPool.stream().map(ReplicaId::getReplicaType).anyMatch(Predicate.isEqual(ReplicaType.CLOUD_BACKED));
     totalReplicaCount = replicaPool.size();
-    if (totalReplicaCount < successTarget) {
-      // {@link MockPartitionId#getReplicaIds} is returning a shared reference which may cause race condition.
-      // Please report the test failure if you run into this exception.
-      throw new IllegalArgumentException(
-          generateErrorMessage(partitionId, examinedReplicas, replicaPool, backupReplicasToCheck, downReplicasToCheck));
+
+    // MockPartitionId.getReplicaIds() is returning a shared reference which may cause race condition.
+    // Please report the test failure if you run into this exception.
+    Supplier<IllegalArgumentException> notEnoughReplicasException = () -> new IllegalArgumentException(
+        generateErrorMessage(partitionId, examinedReplicas, replicaPool, backupReplicasToCheck, downReplicasToCheck));
+    // initialize this to the replica type of the first request to send so that parallelism is set correctly for the
+    // first request
+    inFlightReplicaType =
+        replicaPool.stream().findFirst().map(ReplicaId::getReplicaType).orElseThrow(notEnoughReplicasException);
```

Comment:
If I remember correctly, we are supposed to try replicas in local dc first, then replica in linkedin dcs, and then cloud replica. So this logic has changed?

Reply:
This logic has not changed. The order of the items in replica pool is still the same. This line just gets the replica type of the first item in the linked list, which should be a replica in the local dc.

Reply:
ah, I get your point. The frontend that uses operation tracker might be Azure frontend. In that case, first replica type is cloud replica, right? Did we ensure cloud replica will be picked in first place if when populating the replica pool?

Reply:
The cloud replica will be returned in the list of replica IDs if it exists. Since each cloud replica has a virtual "datanode" it will follow the same datacenter ordering logic that other replicas do. For azure fabrics, it will come first.

Reply:
Make sense. Thanks for the explanation.